### ISSUE TITLE
[BUG] mx.array.tolist does not work for dtype=float16

### ISSUE THREAD

**Issue by arn4** [ml-explore/mlx#701]
**Describe the bug**
`mx.array.tolist` does not work for `dtype=mx.float16`.
I'm trying to do a first naive implementation for #367, and I noticed this bug.

The same things happens with `dtype=mx.bfloat16`

**To Reproduce**

Include code snippet
```python
import mlx.core as mx

a = mx.array([1, 2, 3, 4, 5], dtype=mx.float32)
b = mx.array([1, 2, 3, 4, 5], dtype=mx.float16)

print(a.tolist())
print(b.tolist())
```

It returns 
```
[1.0, 2.0, 3.0, 4.0, 5.0]
Traceback (most recent call last):
  File "/Users/lucaarnaboldi/Desktop/mlx-examples/test.py", line 7, in <module>
    print(b.tolist())
          ^^^^^^^^^^
SystemError: Objects/listobject.c:338: bad argument to internal function
```
**Expected behavior**
It should give the same list twice:
```
[1.0, 2.0, 3.0, 4.0, 5.0]
[1.0, 2.0, 3.0, 4.0, 5.0]
```

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.0]
 - Version [e.g. 0.3.0]

**Comment by awni** [comment]
Thanks for reporting that and the repro. Fix in #702

**Comment by arn4** [comment]
When is merged I can pull request for pickling (#300 & #367)
